bee-trace/bee-trace-attributes/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+/// let location = std::panic::Location::caller();
+///
+/// tracing::trace_span!(
+///     "bee::observe",
+///     "observed",
+///     observed.name = "say_hello",
+///     loc.file = location.file(),
+///     loc.line = location.line(),
+///     loc.col = location.column(),
+/// );
+/// ```
+///
+/// The future or function will then run inside the context of the generated span:
+/// ```ignore
+/// let _guard = span.enter();
+///
+/// async move {
+///     println!("hello");
+/// }
+/// .await;
+/// ```
+#[proc_macro_attribute]
+pub fn observe(_args: proc_macro::TokenStream, input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let observe_impl = parse_macro_input!(input as ObserveImpl);
+    observe_impl.gen_tokens().into()
+}
+
+#[derive(Debug)]
+struct ObserveImpl {
+    attributes: Vec<Attribute>,
+    visibility: Visibility,
+    signature: Signature,
+    block: TokenStream,
+}
+
+impl Parse for ObserveImpl {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let attributes = input.call(Attribute::parse_outer)?;
+        let visibility = input.parse()?;
+        let signature = input.parse()?;
+        let block = input.parse()?;
+
+        Ok(Self {
+            attributes,
+            visibility,
+            signature,
+            block,
+        })
+    }
+}
+
+impl ObserveImpl {
+    fn gen_tokens(self) -> TokenStream {
+        let ObserveImpl {
+            attributes,
+            visibility,
+            signature,
+            block,
+        } = self;
+
+        let fn_name = signature.ident.to_string();
+
+        let block = match &signature.asyncness {
+            Some(_) => Self::gen_async_block(fn_name, &block),
+            None => Self::gen_block(fn_name, &block),
+        };
+
+        quote! {
+            #(#attributes)*
+            #visibility #signature
+            {
+                #block
+            }
+        }
+    }
+
+    fn gen_block(fn_name: String, block: &TokenStream) -> TokenStream {
+        let span = quote! {
+            {
+                let location = std::panic::Location::caller();
+
+                tracing::trace_span!(
+                    target: "bee::observe",
+                    "observed",
+                    observed.name = #fn_name,
+                    loc.file = location.file(),
+                    loc.line = location.line(),
+                    loc.col = location.column(),
+                )
+            }
+        };
+
+        quote_spanned! {
+            block.span() => {
+                let span = #span;
+                let _guard = span.enter();
+                #block
+            }
+        }
+    }
+
+    fn gen_async_block(fn_name: String, block: &TokenStream) -> TokenStream {
+        let observed_future = quote_spanned! {
+            block.span() => async move #block
+        };
+
+        quote! {
+            bee_trace::Observe::observe(#observed_future, #fn_name).await
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
maybe you can use the same test suite setup as `bee-packable-derive` so we can keep tests in separated files. But I'd do that in a different PR. Not sure about the `expected_tokens` part tho.

Reply:
Sounds good, I can do another PR for that later.